src/AutoRest.CSharp/Common/Output/Models/Types/SchemaObjectType.cs

```diff
@@ -565,8 +565,8 @@ property.Type is not InputNullableType &&
             {
                 // In mixed models required properties are not readonly
                 propertyShouldOmitSetter |= property.IsRequired &&
-                              _usage.HasFlag(InputModelTypeUsage.Input) &&
-                              !_usage.HasFlag(InputModelTypeUsage.Output);
+                              !_usage.HasFlag(InputModelTypeUsage.Input) &&
+                              _usage.HasFlag(InputModelTypeUsage.Output);
```

Comment:
in ModelTypeProvider, did we do the same?
Or let me ask in this way:
Are the behaviors of ModelTypeProvider and SchemaObjectType different in the first place? And now we are just fixing them to be the same

Reply:
There is no such logic in ModelTypeProvider, it just takes the readOnly from InputProperty https://github.com/Azure/autorest.csharp/blob/7a6528a102390ace593306bb82025a08ff6465b7/src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs#L525

This tweaking is only for mgmt.

Reply:
well this ctor is taking the `readonly` flag from the `field`, and whether the field has readonly flag comes from here:
https://github.com/Azure/autorest.csharp/blob/7a6528a102390ace593306bb82025a08ff6465b7/src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProviderFields.cs#L204
We should ensure these two pieces generate the same thing or if not, we should know the gap.

Reply:
Also according to Arthur, when generating from swagger, the setter is fine. The issue comes when we change the input to typespec.
As far as I know, when the input is swagger, we use `SchemaObjectType` to generate models, and when the input is typespec, we use `ModelTypeProvider` to generate models.
According to the generated code, swagger-generated code is correct, therefore we should change `ModelTypeProvider` (the code above)

The "safe flatten" part is done in the `ObjectTypeProperty` class therefore it applies to both swagger input and typespec input - it is only enabled by the arm flag.

Reply:
>As far as I know, when the input is swagger, we use SchemaObjectType to generate models, and when the input is typespec, we use ModelTypeProvider to generate models.

Not really. For tsp input, we also use `SchemaObjectType`
https://github.com/Azure/autorest.csharp/blob/7a6528a102390ace593306bb82025a08ff6465b7/src/AutoRest.CSharp/Common/AutoRest/Plugins/CSharpGen.cs#L84-L89

Reply:
The reason the swagger input was correct is because the usage was input | output.
And the usage of tsp input is input only, without output.